src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,275 @@
+                            position,
+                        },
+                        time,
+                    ) in vector.drain(..)
+                    {
+                        let value = values
+                            .entry(cap.delayed(&time))
+                            .or_insert_with(HashMap::new)
+                            .entry(key)
+                            .or_insert_with(Default::default);
+
+                        if let Some(new_offset) = position {
+                            if let Some(offset) = value.1 {
+                                if offset < new_offset {
+                                    *value = (val, position);
+                                }
+                            } else {
+                                *value = (val, position);
+                            }
+                        }
+                    }
+                });
+
+                // Produce (key, val) pairs at any complete times.
+                for (cap, map) in values.iter_mut() {
+                    if !input.frontier.less_equal(cap.time()) {
+                        let mut session = output.session(cap);
+                        for (key, val) in map.drain() {
+                            session.give(((key, val), cap.time().clone()))
+                        }
+                    }
+                }
+                // Discard entries, capabilities for complete times.
+                values.retain(|_cap, map| !map.is_empty());
+            }
+        },
+    )
+}
+
+fn apply_linear_operators<G>(
+    stream: &Stream<G, (Row, Option<Row>, Timestamp)>,
+    operator: &LinearOperator,
+    src_type: &RelationType,
+) -> (
+    Stream<G, (Row, Option<Row>, Timestamp)>,
+    Stream<G, DataflowError>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // Determine replacement values for unused columns.
+    // This is copied over from applying linear operators to
+    // the non-upsert case.
+    // At this point, the stream consists of (key, key+payload, time)
+    // so it is ok to delete replace unused values in key+payload
+    // but not to replace unused values in key because that would cause
+    // errors in arrange_from_upsert
+    let position_or = (0..src_type.arity())
+        .map(|col| {
+            if operator.projection.contains(&col) {
+                Ok(col)
+            } else {
+                Err({
+                    // TODO(frank): This could be `Datum::Null` if we
+                    // are certain that no readers will consult it and
+                    // believe it to be a non-null value. That is the
+                    // intent, but it is not yet clear that we ensure
+                    // this.
+                    let typ = &src_type.column_types[col];
+                    if typ.nullable {
+                        Datum::Null
+                    } else {
+                        typ.scalar_type.dummy_datum()
+                    }
```

Comment:
This will interact/clash with @benesch's PR that introduces `Datum::Dummy`. Should be easy to reconcile logically, but flagging so that no hair is pulled out.

Reply:
Since rebasing would screw with the commit history, would you like to rebase now or after you have had a chance to review it?